### PR TITLE
[`ReanimatedSwipeable`] Multiple bug fixes and improvements

### LINKED ISSUE #3147: ReanimatedSwipable Delayed
### Description

Tried upgrading from `Swipeable` to `ReanimatedSwipable` but had to rollback.

For some reason the `progress` value on `ReanimatedSwipable` updates delayed as to what happened on the UI, the swipeable is already open but the progess.value, onSwipeableOpen, etc all fire 1-2 seconds after. This was not an issue with `Swipeable`

### Steps to reproduce

Test side-by-side a Swipeable and ReanimatedSwipable

### Snack or a link to a repository

N/A

### Gesture Handler version

2.20.0

### React Native version

0.75.4

### Platforms

iOS

### JavaScript runtime

Hermes

### Workflow

Expo bare workflow

### Architecture

Paper (Old Architecture)

### Build type

Debug mode

### Device

iOS simulator

### Device model

iPhone 16 Pro Max - iOS 18

### Acknowledgements

Yes

### PR DESCRIPTION
**Pull request by latekvo**
## Description

This PR applies multiple bug fixes and improvements to the `ReanimatedSwipeable`

### Changes:

- Fix reversed swipe direction being supplied to `open` and `willOpen` callbacks.
- Fix `startDrag` callbacks receiving invalid `direction` when opening swipeable.
- Fix unpredictable swipe direction being supplied to `close` and `willClose` callbacks.
- Fix `onSwipeableWillClose` and `onSwipeableClose` callbacks sometimes being called on open.
- Fix `progress` value skipping last `10%` of it's range and having a delay relative to the `translation`.  
  closes #3147

## Test plan

Will create an example showcasing outputs of all the affected callbacks, work in progress.

### CODE REVIEW

**Comment by m-bert** [comment]
> This PR applies multiple bug fixes and improvements to the ReanimatedSwipeable

Is it necessary to fix all of them in just 1 PR? I'd prefer to split it into smaller ones if possible.

**Comment by latekvo** [comment]
>Is it necessary to fix all of them in just 1 PR? I'd prefer to split it into smaller ones if possible.

Most of them are very minor, most also depend on each other / are fixes to the same part of `ReanimatedSwipeable`.

I can split them up if you'd prefer that but i think it's better to keep them as one PR and use a single unified test for all of them.

File: src/components/ReanimatedSwipeable.tsx [hunk 1]
```diff
@@ -391,17 +396,34 @@ const Swipeable = forwardRef<SwipeableMethods, SwipeableProps>(
...
 
+        const frozenRowState = rowState.value;
+
         appliedTranslation.value = withSpring(
           toValue,
           translationSpringConfig,
           (isFinished) => {
             if (isFinished) {
-              runOnJS(dispatchEndEvents)(fromValue, toValue);
+              runOnJS(dispatchEndEvents)(frozenRowState, toValue);
```
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:426]: Are all events executed on the JS thread?

**Review by m-bert — COMMENTED** [review]
I still think it could be split into smaller PRs (like switching to `enum` instead of `string`), but I won't block it. Please wait for @j-piasecki final opinion on [this](https://github.com/software-mansion/react-native-gesture-handler/pull/3149#discussion_r1799636679)

File: src/components/ReanimatedSwipeable.tsx (hunk 1 above)
Comment (j-piasecki) [reply src/components/ReanimatedSwipeable.tsx:426]: I think we can update the calling logic to check whether the callback is a worklet and run it on the UI thread if it is (that's for another PR though).

**Review by j-piasecki — APPROVED** [review]
I think we can merge this, the changes are not necessarily related but the PR isn't that big. In the future though, please try to open separate PRs for unrelated changes (like extracting direction to enum).

Please address https://github.com/software-mansion/react-native-gesture-handler/pull/3149#discussion_r1804276795 first.

File: src/components/ReanimatedSwipeable.tsx (hunk 1 above)
Comment (latekvo) [reply src/components/ReanimatedSwipeable.tsx:426]: Great idea, will do that.

File: src/components/ReanimatedSwipeable.tsx [hunk 2]
```diff
@@ -197,6 +205,11 @@ export interface SwipeableMethods {
   reset: () => void;
 }
 
+enum SwipeDirection {
+  LEFT = 'left',
+  RIGHT = 'right',
+}
+
```
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/ReanimatedSwipeable.tsx]: done in [a329a6d](https://github.com/software-mansion/react-native-gesture-handler/pull/3149/commits/a329a6d3391039510001cd9dcfb0bb2ae3210484)

**Review by m-bert — APPROVED** [review]

**Pull request merged**